src/compiler.ts

```diff
@@ -7339,6 +7339,14 @@ export class Compiler extends DiagnosticEmitter {
       }
       case ElementKind.Global: {
         let global = <Global>target;
+        if (!global.hasDecorator(DecoratorFlags.Lazy) && global.hasDecorator(DecoratorFlags.Global) && !global.is(CommonFlags.Compiled)) {
+          this.error(
+            DiagnosticCode.Variable_0_used_before_its_declaration,
+            expression.range,
+            global.internalName
+          );
+          return module.unreachable();
+        }
```

Comment:
Does this cover all possible cases? For example, if the global is not read (`a = t1`), but assigned (`t1 = something`), it looks like this branch alone would not suffice.